modules/saku-policy-store/src/saku/schemas.clj

```diff
@@ -1,53 +1,198 @@
+    [malli.util :as mu]))
 
-(def effect
-  (m/schema [:tuple [:= :effect] [:enum :deny :allow]]))
+(defn rename-keys
+  [m kmap]
+  (reduce-kv
+    (fn [m' k-src fn-or-k-target]
+      (let [v-src (get m k-src)
+            [k-target v] (if (fn? fn-or-k-target)
+                           (fn-or-k-target v-src)
+                           [fn-or-k-target v-src])]
+        (-> m'
+          (cond-> v (assoc k-target v))
+          (cond-> (not= k-src k-target) (dissoc k-src)))))
+    m kmap))
 
-(def identity-statement
-  (m/schema [:map
-             [:statement/actions [:sequential :string]]
-             [:statement/effect effect]
-             [:statement/resources [:sequential :string]]]))
-
-(def resource-statement
-  (m/schema [:map
-             [:statement/actions [:sequential :string]]
-             [:statement/effect effect]
-             [:statement/identities [:sequential :string]]]))
+(defn- -compile-rename-keys-transformer
+  [kmap]
+  {:leave (if (seq kmap)
+            #(rename-keys % kmap)
+            identity)})
 
-(def resource-policy
-  (m/schema [:map
-             [:policy/drn :string]
-             [:policy/statements [:sequential {:min 1} resource-statement]]]))
-
-(def resource-policies
-  (m/schema [:sequential resource-policy]))
-
-(def identity-policy
-  (m/schema [:map
-             [:policy/drn :string]
-             [:policy/statements [:sequential {:min 1} identity-statement]]]))
+(defn rename-keys-transformer
+  [{:keys [rename-key]}]
+  (mt/transformer
+    {:decoders
+     {:map
+      {:compile
+       (fn [schema _]
+         (-compile-rename-keys-transformer
+           (into {}
+             (keep (fn [[k-src val-schema]]
+                     (when-let [rename (get (m/properties val-schema) rename-key)]
+                       [k-src rename])))
+             (m/entries schema))))}}}))
 
-(def identity-policies
-  (m/schema [:sequential identity-policy]))
+(def PolicyType (m/schema [:enum "Resource" "Identity"]))
 
-(def evaluate-one-args
+(def EvaluateOneInput
   (m/schema [:map
              [:drn :string]
              [:actionId :string]
              [:identities [:sequential {:min 1} :string]]]))
 
-(def evaluate-many-args
+(def EvaluateManyInput
   (m/schema [:map
              [:drns [:sequential {:min 1} :string]]
              [:actionId :string]
              [:identities [:sequential {:min 1} :string]]]))
 
-(defn assert*
-  ([schema obj]
-   (when-not (m/validate schema obj)
-     (throw (ex-info "Invalid type" {:anomaly/category ::invalid-type
-                                     :schema schema
-                                     :obj obj
-                                     :cause (m/explain schema obj)})))))
+(def StatementInput
+  [:map
+   [:sid {:optional true
+          :output-k :statement/sid} string?]
+   [:actionIds {:output-k :statement/actions} [:sequential string?]]
+   [:effect {:output-k (fn [x] [:statement/effect [:effect (csk/->kebab-case-keyword x)]])} string?]])
```

Comment:
Oh so similar to the TypeScript type definitions. Feels like we could even emit TS types from here 🤪